Improvement/wrap vega lite line chart as a component

ChengYanJin opened the pull request:
**Component**: LineChart 

**Description**:  
We create the wrapper **VegaChart** for Vega-Lite, it's the top layer wrapper. 
Based on **VegaChart** we can specify other wrappers for different kinds of charts, such as **LineChart**, sparkline and piechart.

- Props of the wrappers

   1. VegaChart: id, spec

   2. LineChart: id, data, xAxis, yAxis

- Responsiveness
Set the height/width to 'container' will make the chart responsive but you need to ensure that the container's width or height is determined outside the plot. (You need to add height/width props in the spec)

**Design**:
The reason we chose Vege-Lite as a chart library please find Wiki page.
[link to the Wiki page](https://github.com/scality/core-ui/wiki/Research-of-the-chart-library)

Closes: #107

ChengYanJin committed:
linechart: Add the vega,vega-lite,vega-embed as dev-dependencies and peerDependencies

These dependencies are required by vega-lite

Refs: #107

bert-e commented:
# Hello chengyanjin,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/core-ui/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @chengyanjin

* @rachedbenmustapha

* @nicolas2bert

* @monpote

carlito-scality commented on package.json:
```diff
@@ -82,5 +82,10 @@
       "pre-commit": "npm run lint"
     }
+  },
+  "dependencies": {
```
You should use "peerDependencies" and "devDependencies" like others

carlito-scality commented on src/lib/components/linechart/LineChart.component.js:
```diff
@@ -0,0 +1,50 @@
+  ...rest
+}: Props) {
+  const spec = {
+    $schema: "https://vega.github.io/schema/vega-lite/v4.json",
```
It is mandatory? if it is the case, be careful, it should not work offline.

ChengYanJin replied:
I will remove it.

carlito-scality commented on src/lib/components/linechart/LineChart.component.js:
```diff
@@ -0,0 +1,50 @@
+
+  return (
+    <LineChartContainer
+      id={id.substr(1)}
```
It would be better if "id" without # 
=> ```vegaEmbed(`#${id}`, spec, { renderer: "svg" });```

carlito-scality commented on src/lib/components/linechart/LineChart.component.js:
```diff
@@ -0,0 +1,50 @@
+};
+
+const LineChartContainer = styled.div`
+  width: 100%;
```
width and height are necessary here?

carlito-scality commented on stories/linechart.js:
```diff
@@ -0,0 +1,146 @@
+import { storiesOf } from "@storybook/react";
+import LineChart from "../src/lib/components/linechart/LineChart.component";
+
+const data = {
```
please move them to the stories/data folder

carlito-scality reviewed (changes requested):
Your component looks like a "Wrapper" for all type of charts.
IMO, we should create a component `Chart` as a wrapper for all charts. It is used to transform chart into a React component, so it takes the big `spec` JSON as it is in input.
And we create another component `LineChart` base on `Chart`. This `LineChart` component has simpler inputs that we transform into the the big `spec` JSON.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @chengyanjin

* @rachedbenmustapha

* @nicolas2bert

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @carlito-scality

ChengYanJin committed:
linechart: Update the test

Refs:#107

ChengYanJin committed:
linechart: Wrap the linechart based on VegaChart wrapper

VegaChart is the wrapper for Vega-Lite library, we define different types of chart based on it.

Refs:#107

carlito-scality commented on package.json:
```diff
@@ -59,7 +59,10 @@
+    "vega-embed": "^6.0.0",
+    "vega-lite": "^4.0.0-beta.10"
   },
   "peerDependencies": {
```
``` 
   "vega": "^5.7.3",
    "vega-embed": "^6.0.0",
    "vega-lite": "^4.0.0-beta.10"
```
should be in `peerDependencies`as well

carlito-scality reviewed (approved):
LGTM!

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @chengyanjin

* @rachedbenmustapha

* @nicolas2bert

* @monpote

nicolas2bert commented on src/lib/components/linechart/LineChart.component.js:
```diff
@@ -0,0 +1,48 @@
+  return <VegaChart id={id} spec={spec}></VegaChart>;
+}
+
+export default LineChart;
```
Please, export `LineChart` in `src/lib/index.js`. :)

bert-e commented:
# Conflict




There is a conflict between your branch `improvement/wrap-vega-lite-line-chart-as-a-component` and the
destination branch `development/1.0`.

Please resolve the conflict on **the feature branch** (`improvement/wrap-vega-lite-line-chart-as-a-component`).

```bash
 $ git fetch
 $ git checkout origin/improvement/wrap-vega-lite-line-chart-as-a-component
 $ git merge origin/development/1.0
 $ # <intense conflict resolution>
 $ git commit
 $ git push origin HEAD:improvement/wrap-vega-lite-line-chart-as-a-component
```

nicolas2bert commented on src/lib/components/linechart/LineChart.component.js:
```diff
@@ -0,0 +1,48 @@
+  return <VegaChart id={id} spec={spec}></VegaChart>;
+}
+
+export default LineChart;
```
Please, export `LineChart` component in src/lib/index.js

ChengYanJin replied:
Good catch! Thanks!

nicolas2bert commented on package.json:
```diff
@@ -70,7 +73,10 @@
     "react-virtualized": "9.21.0",
     "react-select": "3.0.3",
-    "styled-components": "^4.1.2"
+    "styled-components": "^4.1.2",
```
If needed, add these 3 peer dependencies to README.md: https://github.com/scality/core-ui#how-to-use-it

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @chengyanjin

* @rachedbenmustapha

* @nicolas2bert

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @nicolas2bert

ChengYanJin committed:
linechart: export VegaChart and LineChart from index and update README

Refs: #107

ChengYanJin commented:
/approve

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @carlito-scality

* @chengyanjin

* @rachedbenmustapha

* @nicolas2bert

* @monpote





The following reviewers are expecting changes from the author, or must review again:

* @nicolas2bert







*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/1.0`




There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/1.0`




Please check the status of the associated issue None.

Goodbye chengyanjin.
